src/macrofier/engine.mjs

```diff
@@ -1140,46 +1214,47 @@ function insertExampleMacros(template, examples, method, json, templates) {
   return template.replace(/\$\{method\.examples\}/g, content)
 }
 
-function generateResult(result, json, templates) {
-  const type = types.getSchemaType(result, json, { destination: state.destination, section: state.section  })
+function generateResult(result, json, templates, { name = '' } = {}) {
+
+  const type = types.getSchemaType(result, json, { name: name, destination: state.destination, section: state.section  })
 
   if (result.type === 'object' && result.properties) {
     let content = getTemplate('/types/object', templates).split('\n')
 
     for (var i=0; i<content.length; i++) {
       if (content[i].indexOf("${property}") >= 0) {
-        content[i] = Object.entries(result.properties).map(([title, property]) => insertSchemaMacros(content[i], title, property, json)).join('\n')
+        content[i] = Object.entries(result.properties).map(([title, property]) => insertSchemaMacros(content[i], title || name, property, json)).join('\n')
       }
     }
 
-    return insertSchemaMacros(content.join('\n'), result.title, result, json)
+    return insertSchemaMacros(content.join('\n'), name, result, json)
   }
   else if (type === 'string' && Array.isArray(result.enum)) {
-    return insertSchemaMacros(getTemplate('/types/enum', templates), result, json)
+    return insertSchemaMacros(getTemplate('/types/enum', templates), name, result, json)
   }
   else if (result.$ref) {
-    const link = getLinkForSchema(result, json)
+    const link = getLinkForSchema(result, json, { name: name})
 
     // if we get a real link use it
     if (link !== '#') {
       return `[${types.getSchemaType(result, json, { destination: state.destination, section: state.section  })}](${link})`
     }
     // otherwise this was a schema with no title, and we'll just copy it here
     else {
-      const schema = localizeDependencies(result, json)
+      const sch = localizeDependencies(result, json)
       return getTemplate('/types/default', templates)
-              .replace(/\$\{type\}/, types.getSchemaShape(schema, json, { name: result.$ref.split("/").pop()}))
+              .replace(/\$\{type\}/, types.getSchemaShape(sch, json, { name: result.$ref.split("/").pop() }))
     }
   }
   else {
-    return insertSchemaMacros(getTemplate('/types/default', templates), result.title, result, json)
+    return insertSchemaMacros(getTemplate('/types/default', templates), name, result, json)
   }
 }
 
 function insertSchemaMacros(template, title, schema, module) {
   return template.replace(/\$\{property\}/g, title)
-          .replace(/\$\{type\}/g, types.getSchemaType(schema, module, { destination: state.destination, section: state.section, code: false }))
-          .replace(/\$\{type.link\}/g, getLinkForSchema(schema, module))
+          .replace(/\$\{type\}/g, types.getSchemaType(schema.schema ? schema.schema : schema, module, { name: title, destination: state.destination, section: state.section, code: false }))
```

Comment:
`insertSchemaMacros` is only called by `generateResult`, which we have updated to always take a real schema as it's parameter, so I think these extra safety checks aren't needed now.

Reply:
updated to use schema itself